// @decorators/mongoose/lib/src/decorators/schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const meta_1 = require("../meta");
/**
 * Defines Schema field
 * @param {*} type Field type
 */
function SchemaField(type) {
    return (target, name) => {
        meta_1.getMongooseMeta(target).schema[name] = type;
    };
}
exports.SchemaField = SchemaField;
/**
 * Defines static method or property
 */
function Static() {
    return (target, name, descriptor) => {
        meta_1.getMongooseMeta(target.prototype).statics.push(descriptor ? [name, target[name]] : name);
    };
}
exports.Static = Static;
/**
 * Defines query method
 */
function Query() {
    return (target, name) => {
        meta_1.getMongooseMeta(target).queries.push([name, target[name]]);
    };
}
exports.Query = Query;
/**
 * Defines instance method
 */
function Instance() {
    return (target, name) => {
        meta_1.getMongooseMeta(target).instances.push([name, target[name]]);
    };
}
exports.Instance = Instance;
/**
 * Defines virtual (computed) property (getter/setter)
 */
function Virtual() {
    return (target, name, descriptor) => {
        meta_1.getMongooseMeta(target).virtuals.push([name, descriptor]);
    };
}
exports.Virtual = Virtual;
/**
 * Defines index
 */
function Index() {
    return (target, name) => {
        meta_1.getMongooseMeta(target).indexes.push(name);
    };
}
exports.Index = Index;
/**
 * Defines pre-hoos
 *
 * @param {string} hookType
 * @param {string} actionType
 */
function Hook(hookType, actionType) {
    return (target, name) => {
        meta_1.getMongooseMeta(target).hooks.push([hookType, actionType, name]);
    };
}
exports.Hook = Hook;
